### ISSUE TITLE
undefined method `variable_identifiers' for #<GraphQL::Language::Nodes::Argument:0x000055865b8741c8>

### ISSUE THREAD

**Issue by 23tux** [rmosolgo/graphql-ruby#3043]
**Describe the bug**

I'm using a `GraphQL::Language::Visitor` to change the name and definition of a variable. Changing the definition works, but changing the variable identifier throws an error.

**Versions**

`graphql` version: 1.10.10
`rails` 5.2.4.3
`graphql-batch` 0.4.2

**GraphQL schema**

*Not needed, happens at query parse time.*

**GraphQL query**

```graphql
query($id: ID!) {
  product(id: $id) {
    name { de }
  }
}
```

**Steps to reproduce**

```
class MyVisitor < GraphQL::Language::Visitor
  def on_variable_identifier(variable_identifier, argument)
    variable_identifier = variable_identifier.merge(name: variable_identifier.name + "foobar")
    argument = argument.merge(value: variable_identifier, children: [variable_identifier])
    super(variable_identifier, argument)
  end
end

query = <<-GQL
query($id: ID!) {
  product(id: $id) {
    name { de }
  }
}
GQL

doc = GraphQL.parse(query)
query = MyVisitor.new(doc).visit.to_query_string
```

**Expected behavior**

The variable should change it's name.

**Actual behavior**

An error is thrown:

```
Unhandled error during GraphQL execution:

  undefined method `variable_identifiers' for #<GraphQL::Language::Nodes::Argument:0x00007f489bfb53c0>
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/nodes.rb:96:in `public_send'
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/nodes.rb:96:in `replace_child'
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/visitor.rb:180:in `on_node_with_modifications'
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/visitor.rb:102:in `block in on_abstract_node'
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/visitor.rb:101:in `each'
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/visitor.rb:101:in `on_abstract_node'
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/visitor.rb:123:in `on_argument'
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/visitor.rb:79:in `public_send'
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/visitor.rb:79:in `visit_node'
    /usr/local/bundle/gems/graphql-1.10.10/lib/graphql/language/visitor.rb:174:in `on_node_with_modifications'
    ... and 178 more lines
```

**Additional context**

I only included the relevant `on_variable_identifier`, I also change the definition using `on_variable_definition`, this works.

**Comment by rmosolgo** [comment]
Thanks for the detailed report. 

What if you try it without the `children: [...]` part? 

```diff
-     argument = argument.merge(value: variable_identifier, children: [variable_identifier])
+     argument = argument.merge(value: variable_identifier)
```

I don't think Nodes::Argument has children, only `name` and `value`: 

https://github.com/rmosolgo/graphql-ruby/blob/a1df05cf467aefc97d96ef4990b917e861807a11/lib/graphql/language/nodes.rb#L287-L288

Does that work?

**Comment by 23tux** [comment]
Thanks for the quick response!

I tried your diff, but get the same error.

When I'm inspecting the argument variable in pry, I get this:

```
    2: def on_variable_identifier(variable_identifier, argument)
    3:   binding.pry
 => 4:   variable_identifier = variable_identifier.merge(name: variable_identifier.name + "foobar")
    5:   argument = argument.merge(value: variable_identifier)
    6:   super(variable_identifier, argument)
    7: end

[1] pry(#<MyVisitor>)> argument
=> #<GraphQL::Language::Nodes::Argument:0x000055e42816eb50
 @children=[#<GraphQL::Language::Nodes::VariableIdentifier:0x000055e42816ee70 @col=15, @filename=nil, @line=2, @name="id">],
 @col=11,
 @filename=nil,
 @line=2,
 @name="id",
 @value=#<GraphQL::Language::Nodes::VariableIdentifier:0x000055e42816ee70 @col=15, @filename=nil, @line=2, @name="id">>
[2] pry(#<MyVisitor>)>
```

Therefore I thought I needed to merge `children` as well, in order to "notify" `argument` about the changed variable identifier.

**Comment by rmosolgo** [comment]
Thanks for giving that a try. I added a test, which failed just like you said, and pushed a fix here: #3045  

By the way, GraphQL-Ruby will update `parent` for you, so you could remove this line:

```ruby
argument = argument.merge(value: variable_identifier)
```

**Comment by 23tux** [comment]
Thanks! Should I close this issue? Or does it get closed automatically, once the fix is merged?
